cpp/src/arrow/compute/kernels/cast.cc

```diff
@@ -980,6 +980,36 @@ struct CastFunctor<O, StringType,
   }
 };
 
+// ----------------------------------------------------------------------
+// String to Timestamp
+
+template <>
+struct CastFunctor<TimestampType, StringType> {
+  void operator()(FunctionContext* ctx, const CastOptions& options,
+                  const ArrayData& input, ArrayData* output) {
+    //     const auto& out_type = checked_cast<const TimestampType&>(*output->type);
+    using out_type = TimestampType::c_type;
+
+    StringArray input_array(input.Copy());
+    auto out_data = output->GetMutableValues<out_type>(1);
+    internal::StringConverter<TimestampType> converter(output->type);
+
+    for (int64_t i = 0; i < input.length; ++i, ++out_data) {
+      if (input_array.IsNull(i)) {
+        continue;
+      }
+
+      auto str = input_array.GetView(i);
+      if (!converter(str.data(), str.length(), out_data)) {
```

Comment:
I think you need to set the is valid bit. 

I would add a way to skip errors (safe/unsafe) as opposed to bork. This become very annoying in huge 
(always broken in some way) datasets.

Reply:
I wouldn't be surprised if there's already a JIRA open for a "nulls on errors" option :-)

Reply:
Some casts do it.

I'd also propose to simplify the whole CastOptions, to

- Toggling stop on errors (safe vs unsafe), call it "allow_truncation".
- Truncation mode: set bit to null _or_ default to a defined value, e.g. `0` or `INT64_MIN`, "NA", etc.

 @wesm what do you think?

Reply:
https://issues.apache.org/jira/browse/ARROW-1489